src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+namespace Microsoft.Rest
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+
+    /// <summary>
+    /// Abstract base class for types that act like enums, but can be extended with arbitrary string values.
+    /// </summary>
+    public abstract class ExtensibleEnum<T> : IEquatable<T> where T : ExtensibleEnum<T>
+    {
+        private static readonly Lazy<Dictionary<string, T>> _nameMap =
+            new Lazy<Dictionary<string, T>>(CreateNameMap, isThreadSafe: true);
+
+        private string _name;
+
+        /// <summary>
+        /// Initializes a new instance of the ExtensibleEnum class.
+        /// </summary>
+        /// <param name="name">The value for this instance of the extensible enumeration.</param>
+        protected ExtensibleEnum(string name)
```

Comment:
`name` should really be `value` since it's what's about the wire here!

Reply:
`value` would have `TVal` and so on

Reply:
Fair point, can you think of a case where `TVal` could be a non-primitive type

Reply:
well, if you think about Swagger, *any* type is actually allowed. In here, I also don't (yet) see any place that depends on `TVal` being primitive, so I think this generalization should automatically allow any type